File: pagmoSharp/pagmoExtensions/Problems/IProblem.cs
using System;

namespace pagmo
{
    /// <summary>
    /// Managed problem contract for user-defined and wrapped pagmo problems.
    /// </summary>
    /// <remarks>
    /// This mirrors pagmo's UDP concept. Implementers should return consistent fitness/bounds dimensions and
    /// accurately report capability flags (<c>has_*</c>) and <see cref="thread_safety"/> for runtime safety.
    /// </remarks>
    public interface IProblem : IDisposable
    {
        /// <summary>
        /// Returns a human-readable problem name.
        /// </summary>
        string get_name() => "C# problem";

        /// <summary>
        /// Returns additional problem metadata for diagnostics/display.
        /// </summary>
        string get_extra_info() => string.Empty;

        /// <summary>
        /// Evaluates the fitness vector at the provided decision vector.
        /// </summary>
        /// <param name="x">Decision vector.</param>
        /// <returns>Fitness vector.</returns>
        DoubleVector fitness(DoubleVector x);

        /// <summary>
        /// Evaluates fitness for a flattened batch of decision vectors when supported.
        /// </summary>
        /// <param name="dvs">Flattened batch decision vector input.</param>
        /// <returns>Flattened batch fitness output.</returns>
        DoubleVector batch_fitness(DoubleVector dvs) => throw new NotSupportedException("batch_fitness() is not implemented.");

        /// <summary>
        /// Returns lower/upper bounds for decision variables.
        /// </summary>
        PairOfDoubleVectors get_bounds();

        /// <summary>
        /// Returns objective-count (<c>f</c>) dimension.
        /// </summary>
        uint get_nobj() => 1;

        /// <summary>
        /// Returns equality-constraint count.
        /// </summary>
        uint get_nec() => 0;

        /// <summary>
        /// Returns inequality-constraint count.
        /// </summary>
        uint get_nic() => 0;

        /// <summary>
        /// Returns number of integer decision variables.
        /// </summary>
        uint get_nix() => 0;

        /// <summary>
        /// Reports whether <see cref="batch_fitness"/> is implemented.
        /// </summary>
        bool has_batch_fitness() => false;

        /// <summary>
        /// Reports whether <see cref="gradient"/> is implemented.
        /// </summary>
        bool has_gradient() => false;

        /// <summary>
        /// Returns gradient values at the provided decision vector when supported.
        /// </summary>
        DoubleVector gradient(DoubleVector x) => throw new NotSupportedException("gradient() is not implemented.");

        /// <summary>
        /// Reports whether <see cref="gradient_sparsity"/> is implemented.
        /// </summary>
        bool has_gradient_sparsity() => false;

        /// <summary>
        /// Returns gradient sparsity pattern when supported.
        /// </summary>
        SparsityPattern gradient_sparsity() => throw new NotSupportedException("gradient_sparsity() is not implemented.");

        /// <summary>
        /// Reports whether <see cref="hessians"/> is implemented.
        /// </summary>
        bool has_hessians() => false;

        /// <summary>
        /// Returns hessian matrices at the provided decision vector when supported.
        /// </summary>
        VectorOfVectorOfDoubles hessians(DoubleVector x) => throw new NotSupportedException("hessians() is not implemented.");

        /// <summary>
        /// Reports whether <see cref="hessians_sparsity"/> is implemented.
        /// </summary>
        bool has_hessians_sparsity() => false;

        /// <summary>
        /// Returns hessian sparsity patterns when supported.
        /// </summary>
        VectorOfSparsityPattern hessians_sparsity() => throw new NotSupportedException("hessians_sparsity() is not implemented.");

        /// <summary>
        /// Sets the random seed when stochastic behavior supports explicit seeding.
        /// </summary>
        void set_seed(uint seed) => throw new NotSupportedException("set_seed() is not implemented.");

        /// <summary>
        /// Reports whether <see cref="set_seed"/> is implemented.
        /// </summary>
        bool has_set_seed() => false;

        /// <summary>
        /// Returns thread-safety capability used by threaded pagmo execution paths.
        /// </summary>
        thread_safety get_thread_safety() => thread_safety.none;
    }
}

File: java/core/src/main/java/io/github/samthegliderpilot/pagmo4j/problems/IProblem.java
package io.github.samthegliderpilot.pagmo4j.problems;

import io.github.samthegliderpilot.pagmo4j.*;

/**
 * Contract for user-defined problems (UDPs) in pagmo4j.
 *
 * <p>Mirrors pagmo's UDP concept. The two required methods are {@link #fitness(DoubleVector)}
 * and {@link #get_bounds()}; all others have defaults that declare the feature absent.
 *
 * <h3>Minimal implementation</h3>
 * <pre>{@code
 * class MyProblem extends ManagedProblemBase {
 *     public DoubleVector fitness(DoubleVector x) { return vec(x.get(0) * x.get(0)); }
 *     public PairOfDoubleVectors get_bounds() { return bounds(new double[]{-5}, new double[]{5}); }
 * }
 * }</pre>
 *
 * <h3>Thread safety</h3>
 * <p>Override {@link #get_thread_safety()} to return {@link ThreadSafety#Basic} if the
 * problem is safe to call concurrently. Alternatively, implement
 * {@link IThreadCloneableProblem} and override {@link ManagedProblemBase#clone()} to
 * return an independent copy — pagmo4j will create per-thread clones automatically.
 */
public interface IProblem extends AutoCloseable {

    /** Returns a human-readable problem name used in logging and diagnostics. */
    default String get_name() { return "Java problem"; }

    /** Returns extra diagnostic information (displayed by pagmo's pretty-printer). */
    default String get_extra_info() { return ""; }

    /**
     * Evaluates the problem at decision vector {@code x} and returns the combined
     * fitness vector {@code [f_0, …, f_{nobj-1}, c_eq_0, …, c_ineq_0, …]}.
     *
     * @param x decision vector; length must equal the number of decision variables
     * @return fitness + constraint values; length = nobj + nec + nic
     */
    DoubleVector fitness(DoubleVector x);

    /**
     * Returns the box bounds as a pair {@code (lower, upper)}, each of length {@code nx}.
     * The number of decision variables is inferred from {@code lower.size()}.
     */
    PairOfDoubleVectors get_bounds();

    /**
     * Evaluates a batch of decision vectors in one call (optional optimisation).
     * Override together with {@link #has_batch_fitness()} returning {@code true}.
     *
     * @param dvs concatenated decision vectors, length = nx × nPoints
     * @return concatenated fitness vectors, length = (nobj + nec + nic) × nPoints
     */
    default DoubleVector batch_fitness(DoubleVector dvs) {
        throw new UnsupportedOperationException("batch_fitness() is not implemented.");
    }

    /** Number of objectives (default 1 — single-objective). */
    default long get_nobj() { return 1L; }

    /** Number of equality constraints (default 0). */
    default long get_nec()  { return 0L; }

    /** Number of inequality constraints (default 0). */
    default long get_nic()  { return 0L; }

    /** Number of integer decision variables, counted from the end of the decision vector (default 0). */
    default long get_nix()  { return 0L; }

    /** Returns {@code true} if {@link #batch_fitness(DoubleVector)} is implemented. */
    default boolean has_batch_fitness()     { return false; }

    /** Returns {@code true} if {@link #gradient(DoubleVector)} is implemented. */
    default boolean has_gradient()          { return false; }

    /** Returns {@code true} if {@link #gradient_sparsity()} is implemented. */
    default boolean has_gradient_sparsity() { return false; }

    /** Returns {@code true} if {@link #hessians(DoubleVector)} is implemented. */
    default boolean has_hessians()          { return false; }

    /** Returns {@code true} if {@link #hessians_sparsity()} is implemented. */
    default boolean has_hessians_sparsity() { return false; }

    /** Returns {@code true} if {@link #set_seed(long)} is implemented. */
    default boolean has_set_seed()          { return false; }

    /**
     * Returns the gradient at {@code x} as a dense or sparse vector.
     * Override together with {@link #has_gradient()} returning {@code true}.
     */
    default DoubleVector gradient(DoubleVector x) {
        throw new UnsupportedOperationException("gradient() is not implemented.");
    }

    /** Returns the gradient sparsity pattern. Override with {@link #has_gradient_sparsity()}. */
    default SparsityPattern gradient_sparsity() {
        throw new UnsupportedOperationException("gradient_sparsity() is not implemented.");
    }

    /** Returns the Hessian matrices at {@code x}. Override with {@link #has_hessians()}. */
    default VectorOfVectorOfDoubles hessians(DoubleVector x) {
        throw new UnsupportedOperationException("hessians() is not implemented.");
    }

    /** Returns Hessian sparsity patterns. Override with {@link #has_hessians_sparsity()}. */
    default VectorOfSparsityPattern hessians_sparsity() {
        throw new UnsupportedOperationException("hessians_sparsity() is not implemented.");
    }

    /**
     * Sets the random seed used by stochastic problems.
     * Override together with {@link #has_set_seed()} returning {@code true}.
     */
    default void set_seed(long seed) {
        throw new UnsupportedOperationException("set_seed() is not implemented.");
    }

    /**
     * Returns the thread-safety level of this problem.
     * Default is {@link ThreadSafety#None}; override to declare concurrent safety
     * or implement {@link IThreadCloneableProblem} for per-thread cloning.
     */
    default ThreadSafety get_thread_safety() { return ThreadSafety.None; }

    default void throwIfNotThreadSafe() {
        if (get_thread_safety() == ThreadSafety.None) {
            String hint = (this instanceof IThreadCloneableProblem)
                ? " Alternatively, override clone() to return a non-null independent copy."
                : "";
            throw new IllegalStateException(
                "Managed problem '" + get_name() + "' must declare ThreadSafety.Basic or " +
                "ThreadSafety.Constant for this threaded path." + hint);
        }
    }

    @Override
    default void close() {}
}
